### PR TITLE
Reintroduce some macOS builds to Travis-CI

### LINKED ISSUE #1072: macOS continuous integration
The last two issues ([1](https://github.com/rust-lang-nursery/rust-bindgen/issues/1004), [2](https://github.com/rust-lang-nursery/rust-bindgen/issues/1067)) with macOS build made it clear that we need macOS CI.

@fitzgen noted that once upon a time there was macOS builds on Travis CI, but since they took very long time they were disabled.

Maybe it is worth to run the "full suite" asynchroniously after the merge?

### PR DESCRIPTION
**Pull request by kulp**
Closes #1072 by running a few critical jobs on macOS, but not the whole matrix.

In my experience with Travis in general in the last few years, there has not been a lack of macOS builders as was the case when https://github.com/rust-lang/rust-bindgen/issues/1072#issuecomment-335859660 was written.

**This PR would increase the number of Travis jobs by four, from ~~37 to 41~~ 31 to 35.**

Along the way I refactored `ci/before_install.sh`, making it capable of handling LLVM 9.0.1+ (including 10.0.0). Some of the changes there are not strictly necessary to get macOS builds working again, so let me know if I should pare down the list of changes.

The most significant change here was to set `LIBCLANG_PATH` instead of only `DYLD_LIBRARY_PATH`-plus-a-copy. Perhaps it would be better not to set `LIBCLANG_PATH` if we could avoid it.

### CODE REVIEW

File: ci/before_install.sh [hunk 1]
```diff
@@ -1,10 +1,5 @@
-set -ex
-pushd ~
-
-# Workaround for Travis CI macOS bug (https://github.com/travis-ci/travis-ci/issues/6307)
-if [ "${TRAVIS_OS_NAME}" == "osx" ]; then
-    rvm get head || true
-fi
+#!/usr/bin/env bash
+set -exuo pipefail
```
Comment (emilio) [inline ci/before_install.sh:2]: Seems like these broke all builds as we use `LD_LIBRARY_PATH` below without it being bound to anything.

Comment (kulp) [reply ci/before_install.sh:2]: Yes, I pushed that right before bed and I have subsequently removed `-u` again.

**Comment by kulp** [comment]
**Ignore this PR until further notice.**

**Comment by kulp** [comment]
Oddly, the [GitHub "check" for the Travis job](https://github.com/rust-lang/rust-bindgen/pull/1810/checks?check_run_id=831873947) has gotten out of sync with [the Travis job's actual completed state](https://travis-ci.com/github/rust-lang/rust-bindgen/builds/174105356). (It is not clear why two Pull Request checks exist at all -- possibly I confused things by force-pushing too quickly.)

Given how long (10+ minutes) two of the macOS jobs take to run, we might want to prune the list or move those jobs earlier to let them overlap more with other jobs to reduce total time consumed.

More generally, the CI runtime is pretty long, which would be one reason to drop support for old `libclang` versions as in #1807.

In any case, this PR is ready for review.

**Comment by kulp** [comment]
@emilio, the CI issues were fixed by rebasing. Consider reviewing this again, at your leisure.

File: ci/before_install.sh [hunk 2]
```diff
@@ -1,63 +1,85 @@
...
 function llvm_download() {
-    export LLVM_VERSION_TRIPLE=`llvm_version_triple ${LLVM_VERSION}`
-    export LLVM=clang+llvm-${LLVM_VERSION_TRIPLE}-$1
+    local base_url=$1
+    local arch=$2
+
+    export LLVM=clang+llvm-${LLVM_VERSION_TRIPLE}-$arch
     export LLVM_DIRECTORY="$HOME/.llvm/${LLVM}"
 
+    local base_url
```
Comment (emilio) [inline ci/before_install.sh:58]: Why is this needed when you have `local base_url="$1"` above?

Comment (kulp) [reply ci/before_install.sh:58]: That was a mistake during refactoring. I will clean it up in a future PR. Good catch.

Comment (kulp) [reply ci/before_install.sh:58]: This will be fixed by 71a916cf8e082cf8691e530594436a1493f9405c if #1856 merges.
